machineconfiguration/v1alpha1/types_pinnedimageset.go

```diff
@@ -0,0 +1,77 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=pinnedimagesets,scope=Cluster
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1713
+// +openshift:file-pattern=0000_80_pinnedimagesetMARKERS.crd.yaml
+// +openshift:enable:FeatureGate=PinnedImages
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+
+// PinnedImageSet describes a set of images that should be pinned by CRI-O and
+// pulled to the nodes which are members of the declared MachineConfigPools.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type PinnedImageSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of this pinned image set.
+	// +kubebuilder:validation:Required
+	Spec PinnedImageSetSpec `json:"spec"`
+}
+
+// PinnedImageSetSpec defines the desired state of a PinnedImageSet.
+type PinnedImageSetSpec struct {
+	// pinnedImages is a list of OCI Image referenced by digest that should be
+	// pinned and pre-loaded by the nodes of a MachineConfigPool.
+	// Translates into a new file inside the /etc/crio/crio.conf.d directory
+	// with content similar to this:
+	//
+	//      pinned_images = [
+	//              "quay.io/openshift-release-dev/ocp-release@sha256:...",
+	//              "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+	//              "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+	//              ...
+	//      ]
+	//
+	// These image references should all be by digest, tags aren't allowed.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=2000
+	// +listType=atomic
+	PinnedImages []PinnedImageRef `json:"pinnedImages"`
+}
+
+// PinnedImageRef is an OCI Image referenced by digest.
+//
+// The format of the image ref is:
+// host[:port][/namespace]/name@sha256:<digest>
+// +kubebuilder:validation:Required
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=447
+// +kubebuilder:validation:XValidation:rule=`self.matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*@sha256:[a-f0-9]{64}$')`,message="The image reference must be in the format host[:port][/namespace]/name@sha256:<digest> with a valid SHA256 digest"
```

Comment:
If we exclude the digest (split on @), then am I right in thinking that the rest of the repository name is a valid URL just without a scheme? I wonder if we can use the URL validation in CEL to our advantage here

```
isURL('https://'+self.split('@')[0])
```
We could also then validate the second half independently
```
self.split('@')[1].matches('^sha256:[a-f0-9]{64}$')
```
We would also want to check that the list when split does have 2 entries

But these breakdowns would potentially provide more specific feedback

Reply:
hmm seems sort of hacky for minimal gain I believe the error message is fairly clear here but I will try to improve it a bit.

```
host[:port][/namespace]/name@sha256:<digest>
```